### PR TITLE
Fixes #15858: Computing dynamic groups is very memory intensive, and can lead to FGC or OOM

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/15858

### CODE REVIEW

File: webapp/sources/ldap-inventory/inventory-api/src/main/scala/com/normation/inventory/domain/AgentTypes.scala [hunk 1]
```diff
@@ -176,44 +180,36 @@ object AgentInfoSerialisation {
       )
   }
 
-  def parseSecurityToken(agentType : AgentType, tokenJson: JValue, tokenDefault : Option[String]) : Box[SecurityToken]= {
-    import net.liftweb.json.compactRender
+  def parseSecurityToken(agentType : AgentType, tokenJson: io.circe.JsonObject, tokenDefault : Option[String]) : Box[SecurityToken]= {
```
Comment (ncharles) [inline webapp/sources/ldap-inventory/inventory-api/src/main/scala/com/normation/inventory/domain/AgentTypes.scala]: i'm pretty sure we could do a decoder here, but with the corner cases, i'm not sure it's really possible

File: webapp/sources/ldap-inventory/inventory-api/src/main/scala/com/normation/inventory/domain/AgentTypes.scala [hunk 2]
```diff
@@ -227,20 +223,39 @@ object AgentInfoSerialisation {
...
-     token <- json \ "securityToken" match {
-                case JObject(json) => parseSecurityToken(agentType, json, optToken)
-                case _             => parseSecurityToken(agentType, JNothing, optToken)
-              }
+      json: Json <- io.circe.parser.parse(s) match {
+                   case Left(failure) => Failure(s"Can not parse agent info: ${failure.getMessage()}")
+                   case Right(json)   => Full(json)
+                }
+      // Ensure json is valid json
+      validJson <- json.asObject.map(Full(_)).getOrElse(Failure(s"Error when trying to parse string as JSON Agent: json is '$s')"))
```
Comment (ncharles) [inline webapp/sources/ldap-inventory/inventory-api/src/main/scala/com/normation/inventory/domain/AgentTypes.scala]: most of this code will probably make sad any real user of circe :/

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
First commit is there only for posterity, as it doesn't improve notably (less than 10% with incomplete measurement), and is really complex and fragile (because of my implementation)

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala [hunk 3]
```diff
@@ -217,6 +217,19 @@ class AcceptedNodesLDAPQueryProcessor(
     } }
   }
 
+  override def processOnlyId(query:Query) : Box[Seq[NodeId]] = {
+    //only keep the one of the form Full(...)
+    queryAndChekNodeId(query, NodeInfoService.nodeInfoAttributes, None).map { seq => seq.flatMap {
+      case QueryResult(nodeEntry, _ , _) =>
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala]: I'm not sure you need to get all these attributes here, they don't seem to be use. 
I believe you can juste request `NODE_UUID`

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala [hunk 4]
```diff
@@ -217,6 +217,19 @@ class AcceptedNodesLDAPQueryProcessor(
...
+        processor.ldapMapper.entryToNode(nodeEntry) match {
+          case Full(node) => Seq(node.id)
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala]: Here, I'm not sure why you need to mapping ? You only want the `NodeId`, and the mapping is not really the goad of that method. Bad nodes would appear elsewhere (generation for ex, or node details, etc). 
I believe you can earn quite a lot more memory here by just usgin `nodeDit.NODES.NODE.idFromDn(nodeEntry.dn)`

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/QueryProcessor.scala [hunk 5]
```diff
@@ -53,6 +53,12 @@ trait QueryProcessor {
    */
   def process(query:Query) : Box[Seq[NodeInfo]]
 
+  /**
+   * Only get the nodeId, to update dynamic groups
+   * @param query
+   * @return
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/QueryProcessor.scala]: Can you remove the two param/return comment, they are useless.
Also for the comment, perhaps something like: 
```
Only get node ids corresponding to that request, with minimal consistency check. This method is usefull to maximize performance (low memory, high throughout) for ex for dynamic groups. 
```

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/QueryProcessor.scala [hunk 6]
```diff
@@ -53,6 +53,12 @@ trait QueryProcessor {
...
+   */
+  def processOnlyId(query:Query) : Box[Seq[NodeId]]
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/QueryProcessor.scala]: Perhaps the method should return a `Set[NodeId]` ?

**Comment by fanf** [comment]
This seems to be a very good PR. Can you remove the json/circe commit, since it means a new dep and ++ complexity for almost no gain compared to the other parts of the PR ?

**Comment by ncharles** [comment]
Commit modified

**Comment by fanf** [comment]
OK, squash merging this PR

**Pull request merged**